dune/iga/io/igarefinedgeometries.hh

```diff
@@ -0,0 +1,132 @@
+// SPDX-FileCopyrightText: 2023 The dune-iga developers mueller@ibb.uni-stuttgart.de
+// SPDX-License-Identifier: LGPL-3.0-or-later
+
+#pragma once
+
+#include <dune/geometry/multilineargeometry.hh>
+#include <dune/geometry/virtualrefinement.hh>
+
+namespace Dune::IGA {
+
+  template <class GridView>
+  class IGARefinedGeometries {
+   public:
+    static constexpr int dim = GridView::dimension;
+
+    using Element = MultiLinearGeometry<double, dim, dim>;
+    using Point   = Dune::FieldVector<double, dim>;
+    using Index   = std::uint64_t;
+
+   private:
+    struct ElementData {
+      std::vector<Element> elements;
+      std::vector<Point> vertices;
+      std::vector<Index> indices;
+    };
+
+    std::unordered_map<Index, ElementData> trimmedElementData_;
+    ElementData cubeData{};
+
+   public:
+    IGARefinedGeometries(const GridView& gridView, const int subSampleFull, const int subSampleTrimmed) {
+      assert(subSampleFull >= 0 and subSampleTrimmed >= 0 && "subSamples have to be zero or positive");
+
+      createCubeRefinement(subSampleFull);
+
+      // Obtain trimmed Element Data
+      const auto& indexSet = gridView.indexSet();
+
+      for (auto& element : elements(gridView))
+        if (element.impl().isTrimmed()) {
+          auto [ele, vert, ind] = element.impl().elementSubGrid()->createRefinedGrid(subSampleTrimmed);
+
+          trimmedElementData_.emplace(indexSet.index(element), ElementData{ele, vert, ind});
+        }
+    }
+
+    [[nodiscard]] const std::vector<Element>& getElements(Index eIndex) const {
+      if (isTrimmed(eIndex))
+        return trimmedElementData_.at(eIndex).elements;
+      else
+        return cubeData.elements;
+    }
+
+    [[nodiscard]] const std::vector<Point>& getVertices(Index eIndex) const {
+      if (isTrimmed(eIndex))
+        return trimmedElementData_.at(eIndex).vertices;
+      else
+        return cubeData.vertices;
+    }
+
+    [[nodiscard]] const std::vector<Index>& getIndices(Index eIndex) const {
+      if (isTrimmed(eIndex))
+        return trimmedElementData_.at(eIndex).indices;
+      else
+        return cubeData.indices;
+    }
+
+    [[nodiscard]] Index vertexSubIndex(Index eIndex, Index subEleIndex, Index subEntityIndex) const {
+      if (isTrimmed(eIndex))
+        return trimmedElementData_.at(eIndex).indices[subEleIndex * 3 + subEntityIndex];
```

Comment:
Can you write a comment here that  the 3 and 4 stem from the choice to only allow triangles and quads and that these numbers are their vertex counts? (also indicate this through variable names